src/ripple/app/tx/impl/AMMCreate.cpp

```diff
@@ -0,0 +1,374 @@
+        return tecINSUF_RESERVE_LINE;
+    }
+
+    auto insufficientBalance = [&](STAmount const& asset) {
+        if (isXRP(asset))
+            return xrpBalance < asset;
+        return accountID != asset.issue().account &&
+            accountHolds(
+                ctx.view,
+                accountID,
+                asset.issue(),
+                FreezeHandling::fhZERO_IF_FROZEN,
+                ctx.j) < asset;
+    };
+
+    if (insufficientBalance(amount) || insufficientBalance(amount2))
+    {
+        JLOG(ctx.j.debug())
+            << "AMM Instance: insufficient funds, " << amount << " " << amount2;
+        return tecUNFUNDED_AMM;
+    }
+
+    auto isLPToken = [&](STAmount const& amount) -> bool {
+        if (auto const sle =
+                ctx.view.read(keylet::account(amount.issue().account)))
+            return (sle->getFlags() & lsfAMM);
+        return false;
+    };
+
+    if (isLPToken(amount) || isLPToken(amount2))
+    {
+        JLOG(ctx.j.debug()) << "AMM Instance: can't create with LPTokens "
+                            << amount << " " << amount2;
+        return tecAMM_INVALID_TOKENS;
+    }
+
+    return tesSUCCESS;
+}
+
+static std::pair<TER, bool>
+applyCreate(
+    ApplyContext& ctx_,
+    Sandbox& sb,
+    AccountID const& account_,
+    beast::Journal j_)
+{
+    auto const amount = ctx_.tx[sfAmount];
+    auto const amount2 = ctx_.tx[sfAmount2];
+
+    auto const ammKeylet = keylet::amm(amount.issue(), amount2.issue());
+
+    // Mitigate same account exists possibility
+    auto const ammAccount = [&]() -> Expected<AccountID, TER> {
+        std::uint16_t constexpr maxAccountAttempts = 256;
+        for (auto p = 0; p < maxAccountAttempts; ++p)
+        {
+            auto const ammAccount =
+                ammAccountID(p, sb.info().parentHash, ammKeylet.key);
+            if (!sb.read(keylet::account(ammAccount)))
+                return ammAccount;
+        }
+        return Unexpected(tecDUPLICATE);
+    }();
+
+    // AMM account already exists (should not happen)
+    if (!ammAccount)
+    {
+        JLOG(j_.error()) << "AMM Instance: AMM already exists.";
+        return {ammAccount.error(), false};
+    }
+
+    // LP Token already exists. (should not happen)
+    auto const lptIss = ammLPTIssue(
+        amount.issue().currency, amount2.issue().currency, *ammAccount);
+    if (sb.read(keylet::line(*ammAccount, lptIss)))
+    {
+        JLOG(j_.error()) << "AMM Instance: LP Token already exists.";
+        return {tecDUPLICATE, false};
+    }
+
+    // Create AMM Root Account.
+    auto sleAMMRoot = std::make_shared<SLE>(keylet::account(*ammAccount));
+    sleAMMRoot->setAccountID(sfAccount, *ammAccount);
+    sleAMMRoot->setFieldAmount(sfBalance, STAmount{});
+    std::uint32_t const seqno{
+        ctx_.view().rules().enabled(featureDeletableAccounts)
+            ? ctx_.view().seq()
+            : 1};
+    sleAMMRoot->setFieldU32(sfSequence, seqno);
```

Comment:
Unless AMM accounts are deletable this may not be needed.

Reply:
Thanks, will delete.

Reply:
Actually, if I delete this then I'm getting `FTL:LedgerConsensus Invariant failed: account created with wrong starting sequence number`.